homeassistant/components/mikrotik/__init__.py

```diff
@@ -1,201 +1,46 @@
-"""The mikrotik component."""
+"""The Mikrotik component."""
 import logging
-import ssl
 
-import voluptuous as vol
-import librouteros
-from librouteros.login import login_plain, login_token
-
-from homeassistant.const import (
-    CONF_HOST,
-    CONF_PASSWORD,
-    CONF_USERNAME,
-    CONF_PORT,
-    CONF_SSL,
-    CONF_METHOD,
-)
-from homeassistant.helpers import config_validation as cv
-from homeassistant.helpers.discovery import load_platform
-from homeassistant.components.device_tracker import DOMAIN as DEVICE_TRACKER
-from .const import (
-    NAME,
-    DOMAIN,
-    HOSTS,
-    MTK_LOGIN_PLAIN,
-    MTK_LOGIN_TOKEN,
-    DEFAULT_ENCODING,
-    IDENTITY,
-    CONF_TRACK_DEVICES,
-    CONF_ENCODING,
-    CONF_ARP_PING,
-    CONF_LOGIN_METHOD,
-    MIKROTIK_SERVICES,
-)
+from homeassistant.const import CONF_HOST
+from .const import ATTR_MANUFACTURER, DOMAIN
+from .hub import MikrotikHub
 
 _LOGGER = logging.getLogger(__name__)
 
-MTK_DEFAULT_API_PORT = "8728"
-MTK_DEFAULT_API_SSL_PORT = "8729"
-
-MIKROTIK_SCHEMA = vol.All(
-    vol.Schema(
-        {
-            vol.Required(CONF_HOST): cv.string,
-            vol.Required(CONF_USERNAME): cv.string,
-            vol.Required(CONF_PASSWORD): cv.string,
-            vol.Optional(CONF_METHOD): cv.string,
-            vol.Optional(CONF_LOGIN_METHOD): vol.Any(MTK_LOGIN_PLAIN, MTK_LOGIN_TOKEN),
-            vol.Optional(CONF_PORT): cv.port,
-            vol.Optional(CONF_SSL, default=False): cv.boolean,
-            vol.Optional(CONF_ENCODING, default=DEFAULT_ENCODING): cv.string,
-            vol.Optional(CONF_TRACK_DEVICES, default=True): cv.boolean,
-            vol.Optional(CONF_ARP_PING, default=False): cv.boolean,
-        }
-    )
-)
 
-CONFIG_SCHEMA = vol.Schema(
```

Comment:
We shouldn't totally remove support for config yaml. If we replace `CONF_TRACK_DEVICES` with a system level option to disable entities, that option can be removed.

Reply:
Are you refer to the default system option `disable_new_entities`?
So we should have this disabled by default and only enabled by the user?

Reply:
What's the best user experience? The user will loose the ability to control tracking with known_devices.yaml. The user can now instead control this with the entity registry and the system options for disabling new entities.

If we turn off new entities by default, the user will have to go and turn it on, and then disable entities once the corresponding device connects, and isn't wanted.

We're currently lacking a way to specifically add certain discovered devices. We can only exclude after they have been added.

Reply:
Ok I will do it like this, I will add the import from config yaml so that the user can choose to disable tracking devices before starting HA. And in case it is configured through config flow devices will enabled by default as the user expects to see discovered devices when the configuration is done. They can later control this using entity registry.

Reply:
I don't think the user can have the system level option for a config entry set to disabled and specifically enable entities in the entity registry. It would be good to confirm this.

Reply:
The `disable_new_entities` system option is used for handling new entities and is different from an option to `track_devices` in general (new and existing).
I added the `track_devices` as an option because i believe it is not possible to create it as a system option. How do you recommend the implementation should be?
Should the user have the `track_devices` option as mentioned above or only have control of enabling\disabling on entity level?

Reply:
If the user disables an entity and reloads, the entity will not be added and thus no longer tracked. So it makes the behavior more or less the same.

But we should test this thoroughly so we so that the user experience is satisfactory.

Reply:
I did some testing and i was able to achieve the following:
- If the integration is configured from config flow the detected devices are added but disabled. The user will need to enable each device manually. This will enable the tracking of the device even after HA reloads.
- If the integration is imported from config yaml the system_option reflect the `track_devices` option configured (Defaults to False).
- Disabling an entity will disable its tracking even after a reload.

I would suggest to have the `track_devices` option part of the initial config flow setup so that the user can enable this from the beginning if he wants to. what do you think?

Reply:
When you say that devices are added but disabled, you mean that they are added to the entity registry, right, and not to the state machine?

Why do we need the extra `track_devices` option when we have the system option `disable_new_entities`? Is it to keep support for changing this behavior via config yaml?

Reply:
If the system option `disable_new_entities` is true, can the user still enable an enity in the entity registry and have it be added to the state machine?

Reply:
Yes correct. They are added but disabled (can be enabled in entity registry).

If `disable_new_entities` system option is True initially then all detected devices will be added but disabled. The user will have to enable each entity to track it (and add it to the state machine.

The user might not want to enable each device separately so if he wants to enable tracking by default we need to have this option during the config flow setup and not after.

Reply:
There's a PR where this is discussed:
https://github.com/home-assistant/home-assistant/pull/27044

We should check the system level option in the platform and try to add new entities on every scan so that a change in the system option changes the behavior immediately without reload necessary.

Can you confirm that the user can enable an entity in the entity registry and have it be added to the state machine even while the system option `disable_new_entities` is true?

Reply:
This is the test i did:
1. Configure a Mikrotik integration using config flow. Initially all devices are added with the device_tracker entity showing as `unavailable` and not showing in HA states.
2. Enable an entity from Entity registry. The entity become available on the next update and showing under states.
3. Restart HA --> Only the enabled entity is showing in states and the others are still `unavailable` as expected.
4. Enable system option to allow adding new entities --> New detected devices are added and entities are enabled.
5. Restart HA with system option enabled --> only previously `available` entities are showing in states. Other entities remain `unavailable`.

Reply:
Good!